browser/modules/PlacesProvider.jsm

```diff
@@ -0,0 +1,258 @@
+      // The implementation of the query in getLinks excludes hidden and
+      // unvisited pages, so it's important to exclude them here, too.
+      if (!aHidden && aLastVisitDate) {
+        Links.eventEmitter.emit("linkChanged", {
+          url: aURI.spec,
+          frecency: aNewFrecency,
+          lastVisitDate: aLastVisitDate,
+          type: "history",
+        });
+      }
+    },
+
+    onManyFrecenciesChanged: function PlacesProvider_onManyFrecenciesChanged() {
+      Links.eventEmitter.emit("manyLinksChanged");
+    },
+
+    onTitleChanged: function PlacesProvider_onTitleChanged(aURI, aNewTitle) {
+      Links.eventEmitter.emit("linkChanged", {
+        url: aURI.spec,
+        title: aNewTitle
+      });
+    },
+
+    QueryInterface: XPCOMUtils.generateQI([Ci.nsINavHistoryObserver,
+                        Ci.nsISupportsWeakReference])
+  },
+
+  /**
+   * Must be called before the provider is used.
+   */
+  init: function PlacesProvider_init() {
+    PlacesUtils.history.addObserver(this.historyObserver, true);
+  },
+
+  /**
+   * Gets the current set of links delivered by this provider.
+   *
+   * @returns {Promise} Returns a promise with the array of links as payload.
+   */
+  getLinks: function PlacesProvider_getLinks() {
+    let getLinksPromise = new Promise((resolve, reject) => {
+      let options = PlacesUtils.history.getNewQueryOptions();
+      options.maxResults = this.maxNumLinks;
+
+      // Sort by frecency, descending.
+      options.sortingMode = Ci.nsINavHistoryQueryOptions
+        .SORT_BY_FRECENCY_DESCENDING;
+
+      let links = [];
+
+      let callback = {
+        handleResult: function(aResultSet) {
+          let row;
+
+          while ((row = aResultSet.getNextRow())) {
+            let url = row.getResultByIndex(1);
+            if (LinkChecker.checkLoadURI(url)) {
+              let title = row.getResultByIndex(2);
+              let frecency = row.getResultByIndex(12);
+              let lastVisitDate = row.getResultByIndex(5);
+              links.push({
+                url: url,
+                title: title,
+                frecency: frecency,
+                lastVisitDate: lastVisitDate,
+                type: "history",
+              });
+            }
+          }
+        },
+
+        handleError: function(aError) {
+          reject(aError);
+        },
+
+        handleCompletion: function(aReason) { // jshint ignore:line
+          // The Places query breaks ties in frecency by place ID descending, but
+          // that's different from how Links.compareLinks breaks ties, because
+          // compareLinks doesn't have access to place IDs.  It's very important
+          // that the initial list of links is sorted in the same order imposed by
+          // compareLinks, because Links uses compareLinks to perform binary
+          // searches on the list.  So, ensure the list is so ordered.
+          let i = 1;
```

Comment:
if we are sorting a list we should just call `links.sort(LinkUtils.compareLinks)` and make sure `compareLinks` does the sorting work for us?